### PR TITLE
Introduce Redeems Reserve

### PR DESCRIPTION
**Pull request by rkolpakov**

### CODE REVIEW

File: test/integration/core/redeems-reserve-scenarios.integration.ts [hunk 1]
```diff
@@ -0,0 +1,1129 @@
...
+    });
+  });
+
+  // ═══════════════════════════════════════════════════════════════════════
+  //  Smoothing tightening after refSlot redemption
+  // ═══════════════════════════════════════════════════════════════════════
+
+  context("Smoothing tightening after refSlot redemption", () => {
+    /**
+     * @notice Failure: redemption between refSlot and report reduces smoothing base.
```
Comment (Psirex) [inline test/integration/core/redeems-reserve-scenarios.integration.ts]: Maybe worth using "Edge Case" instead of "Failure". As scenario is not failure

File: test/integration/core/redeems-reserve-scenarios.integration.ts [hunk 2]
```diff
@@ -0,0 +1,1129 @@
...
+      const elRewards = ether("100"); // large enough to hit smoothing limit
+      await updateBalance(elVaultAddr, elRewards);
+
+      const totalPooledBeforeA = await lido.getTotalPooledEther();
+      await doReport({ excludeVaultsBalances: false, reportElVault: true });
+      const totalPooledAfterA = await lido.getTotalPooledEther();
+      const distributedA = totalPooledAfterA - totalPooledBeforeA;
+
+      // Restore to same state
+      await Snapshot.restore(testSnapshot);
```
Comment (Psirex) [inline test/integration/core/redeems-reserve-scenarios.integration.ts]: Do not modify test's snapshot. Instead took the new one local for the test

File: test/integration/redeems-reserve/helpers.ts [hunk 3]
```diff
@@ -0,0 +1,550 @@
...
+}
+
+/** Deposits ETH, sets reserve ratio, and runs a report to fill the reserve */
+export async function seedReserve(
+  ctx: ProtocolContext,
+  holder: HardhatEthersSigner,
+  reserveManager: HardhatEthersSigner,
+  opts: {
+    deposit: bigint;
+    ratioBP: bigint;
```
Comment (Psirex) [inline test/integration/redeems-reserve/helpers.ts]: Consider naming it `redeemsReserveTarget`, similar to `depositsReserveTarget`

File: test/integration/redeems-reserve/helpers.ts [hunk 4]
```diff
@@ -0,0 +1,550 @@
...
+
+/** Computes expected reserve target from internal ether and ratio */
+export function expectedReserveTarget(internalEther: bigint, ratioBP: bigint): bigint {
+  return ratioBP * internalEther / 10_000n;
+}
+
+/** Validates reserve-related state derived from total pooled ether */
+export function assertReserveState(state: ProtocolState, ratioBP: bigint) {
+  expect(state.reserveTarget).to.equal(expectedReserveTarget(state.internalEther, ratioBP));
+  expect(state.reserve).to.equal(state.reserveTarget);
```
Comment (Psirex) [inline test/integration/redeems-reserve/helpers.ts:295]: Seems like it works only when the reserve wasn't used and buffer has enough eth

File: test/integration/redeems-reserve/buffer-allocation.integration.ts [hunk 5]
```diff
@@ -0,0 +1,207 @@
...
+    });
+    await lido.connect(secondHolder).submit(ZeroAddress, { value: SECOND_DEPOSIT });
+    await doReport(ctx, { excludeVaultsBalances: true, skipWithdrawals: true });
+
+    const protocol0 = await captureState(lido);
+
+    await assertReserveAllocationInvariant(lido);
+    assertReserveState(protocol0, DEFAULT_RATIO_BP);
+
+    // --- Request withdrawals exceeding unreserved buffer ---
```
Comment (Psirex) [inline test/integration/redeems-reserve/buffer-allocation.integration.ts:125]: Consider adding an assert which checks that withdrawable ether actually less than the total withdrawals amount

File: test/integration/redeems-reserve/buffer-allocation.integration.ts [hunk 6]
```diff
@@ -0,0 +1,207 @@
...
+
+    const requestAmount = await lido.balanceOf(holder.address);
+    await requestWithdrawals({ ctx, from: holder, amounts: [requestAmount] });
+
+    const beforeReport = await captureBufferState(ctx);
+    const restorePoint = await Snapshot.take();
+    const sharedAllocationBefore = beforeReport.withdrawalsReserve + getUnreserved(beforeReport);
+
+    // --- Path A: report with growthShare = 0 ---
+    await doReport(ctx, { skipWithdrawals: true, excludeVaultsBalances: true });
```
Comment (Psirex) [inline test/integration/redeems-reserve/buffer-allocation.integration.ts:180]: Maybe it would make sense to simulate rewards in this test, to verify that `withdrawalsReserve` grows while the redeem reserve stays unchanged after the report.

File: test/integration/redeems-reserve/helpers.ts [hunk 7]
```diff
@@ -0,0 +1,550 @@
...
+}
+
+/** Funds the EL rewards vault with the given amount */
+export async function fundElRewards(ctx: ProtocolContext, amount: bigint) {
+  const elVaultAddr = await ctx.contracts.locator.elRewardsVault();
+  await updateBalance(elVaultAddr, amount);
+}
+
+/** Advances chain time past the oracle requestTimestampMargin */
+export async function advanceToReportableTime(ctx: ProtocolContext) {
```
Comment (Psirex) [inline test/integration/redeems-reserve/helpers.ts]: Nitpick: the naming is a bit misleading. It suggests that this helper advances time to the point where the oracle can report, while it actually advances time to the point where all withdrawal requests satisfy the creation-time sanity check.

File: test/integration/redeems-reserve/negative-rebase.integration.ts [hunk 8]
```diff
@@ -0,0 +1,244 @@
...
+    await doReport(ctx, { clDiff: CL_LOSS });
+    const state2: ProtocolState = await captureState(lido);
+    await assertReserveAllocationInvariant(lido);
+
+    // --- Compare paths: reserve auto-shrinks to new target in both ---
+    assertReserveState(state2, RATIO_BP);
+
+    // Verify: reserve target matches internalEther-based calculation in both paths
+    expect(state2.reserveTarget).to.equal(expectedReserveTarget(state2.internalEther, RATIO_BP));
+    expect(state1.reserveTarget).to.equal(expectedReserveTarget(state1.internalEther, RATIO_BP));
```
Comment (Psirex) [inline test/integration/redeems-reserve/negative-rebase.integration.ts:115]: Should `state2.reserveTarget` be slightly lower than `state1.reserveTarget`, given that the ratio is now applied to a smaller base? If so, it may be worth adding an assertion to make that behavior explicit.

File: test/integration/redeems-reserve/negative-rebase.integration.ts [hunk 9]
```diff
@@ -0,0 +1,244 @@
...
+    await Snapshot.restore(simSnapshot);
+
+    // --- Path B: small redeem, then report with same CL loss ---
+    const REDEEM_AMOUNT = await getRedeemAmount(lido, "small");
+    const redeemShares = await lido.getSharesByPooledEth(REDEEM_AMOUNT);
+    const redeemEther = await lido.getPooledEthByShares(redeemShares);
+
+    await redeemExact(lido, holder, fix, REDEEM_AMOUNT);
+
+    await doReport(ctx, { clDiff: CL_LOSS });
```
Comment (Psirex) [inline test/integration/redeems-reserve/negative-rebase.integration.ts:106]: Consider adding an assertion to verify that the share rate decreased, confirming that the report was indeed negative

File: test/integration/redeems-reserve/negative-rebase.integration.ts [hunk 10]
```diff
@@ -0,0 +1,244 @@
...
+    const externalShares0 = await lido.getExternalShares();
+    const internalShares0 = state0.totalShares - externalShares0;
+    const quote0 = await captureRedeemQuote(lido, QUOTE_STETH_AMOUNT);
+    await assertReserveAllocationInvariant(lido);
+    assertReserveState(state0, RATIO_BP);
+
+    // --- Internalize bad debt, process report ---
+    const badDebtShares =
+      (await dashboard.liabilityShares()) - (await lido.getSharesByPooledEth(await dashboard.totalValue()));
+    const internalizedBadDebtShares = await vaultHub
```
Comment (Psirex) [inline test/integration/redeems-reserve/negative-rebase.integration.ts:209]: Consider adding check, `internalizedBadDebtShares` is not zero

File: test/integration/redeems-reserve/low-tvl.integration.ts [hunk 11]
```diff
@@ -0,0 +1,153 @@
...
+
+  afterEach(async () => {
+    await Snapshot.restore(testSnapshot);
+  });
+
+  after(async () => {
+    await Snapshot.restore(snapshot);
+  });
+
+  it("small TVL: redeem + WQ + rebase with two holders", async () => {
```
Comment (Psirex) [inline test/integration/redeems-reserve/low-tvl.integration.ts]: Consider adding an additional edge case where withdrawal finalization gets effectively clogged by the redeem reserve.

Example scenario:
Users A, B, and C hold most of the protocol stake. User A withdraws their full position. User B redeems part of their position, consuming the entire redeem reserve, and withdraws the remainder via the Withdrawal Queue. User C then withdraws.

Result: User C cannot be finalized. Even though the reserve has decreased, it is still not enough to cover the full withdrawal request.

Resolution: submit an oracle report that allocates all buffered ether to finalization, which resolves the issue.

File: test/integration/redeems-reserve/redeem-between-ref-and-report.integration.ts [hunk 12]
```diff
@@ -0,0 +1,363 @@
...
+    const redeemShares = await lido.getSharesByPooledEth(redeemAmount);
+    const redeemEther = await lido.getPooledEthByShares(redeemShares);
+    await redeemExact(lido, holder, fix, redeemAmount);
+
+    const headroomDrained = await getHeadroom(lido);
+
+    await doReport(ctx, { excludeVaultsBalances: false, reportElVault: true });
+
+    const state2: ProtocolState = await captureState(lido);
+    const deferredRewards = await ethers.provider.getBalance(elVaultAddr);
```
Comment (Psirex) [inline test/integration/redeems-reserve/redeem-between-ref-and-report.integration.ts:164]: Needed a check, that deferred rewards are not zero

File: test/integration/redeems-reserve/redeem-between-ref-and-report.integration.ts [hunk 13]
```diff
@@ -0,0 +1,363 @@
...
+    await assertReserveAllocationInvariant(lido);
+
+    // --- Fund EL rewards exactly at headroom ---
+    const headroomFull = await getHeadroom(lido);
+    await fundElRewards(ctx, headroomFull);
+
+    // --- Path A: report without redeem (via snapshot) ---
+    const simSnapshot = await Snapshot.take();
+
+    await doReport(ctx, { excludeVaultsBalances: false, reportElVault: true });
```
Comment (Psirex) [inline test/integration/redeems-reserve/redeem-between-ref-and-report.integration.ts:136]: It seems the current implementation of `accounting.report()` always calls `simulateReport()` internally. That may violate the requirement that redeem happens after refSlot. Worth double-checking that, in this test, `simulatedShareRate` is actually calculated from the pre-redeem state.

File: test/integration/redeems-reserve/full-circle.integration.ts [hunk 14]
```diff
@@ -0,0 +1,262 @@
...
+    const batchesState = await withdrawalQueue.calculateFinalizationBatches(simulatedShareRate, maxTimestamp, 1000n, {
+      remainingEthBudget: buffered,
+      finished: false,
+      batches: Array(36).fill(0n),
+      batchesLength: 0n,
+    });
+    const overrideBatches = [...batchesState.batches].filter((x) => x > 0n);
+
+    // --- Process report with override batches ---
+    await report(ctx, {
```
Comment (Psirex) [inline test/integration/redeems-reserve/full-circle.integration.ts:239]: Consider adding an assertion for the redeem reserve: before the report, it should match the corresponding share of the protocol, and after the report, it should be zero.

File: test/integration/redeems-reserve/helpers.ts (hunk 4 above)
Comment (rkolpakov) [reply test/integration/redeems-reserve/helpers.ts:295]: Yes, in cases where reserve drained/not replenished it uses explicit check
